docs/settings.md
# Conversion Settings

### Format
The number format in the output.
- HEX (Example: 0x3c, 0x21, 0x44, 0x4f, 0x43, 0x54, 0x59, ...)
- DEC (Example:  60,  33,  68,  79,  67,  84,  89, ...)
- BIN (Example: b00111100, b00100001, b01000100, b01001111, b01000011, b01010100, b01011001, ...)

### Prefix
What text you want add to the number.  
In C and as HEX, it is `0x` (0x5A), but you can also set '\x' or leave it empty.

### Suffix
What text you want add after the number.  
For example in VB and HEX, you need the `h` after the hex number (5Ah)

### Pad
If you want fill the numbers with '0' (or ' ' for DEC) to have them aligned.  
For example a 0x5 will be converted to 0x05 with 8-bit numbers and to 0x0005 with 16-bit numbers

### Datatype
 - uint8_t for a 8 bit array
 - uint16_t for a 16 bit array
 - uint32_t for a 32 bit array
Array size is updated automatically. You can set it only if the file is divisible by this data type length.

### Columns
The quantity of numbers for each row. Set to 0 if you want all data on a single line.

### Skip Bytes
Bytes to skip. Sometimes you have some headers to remove. If so, set the number of bytes you want to remove from the file, before the array is generated.

### ASCII Comments
Add a column with the converted number as ASCII (helpful if you converts some text documents).


docs/imagesettings.md
# Image Settings
If you import an image, another settings box will appear. So you can do some image manipulations before create the binary array.

## Convert To
### none
Image is converted like any other file
### jpeg
Image is converted to a JPEG before converting it
### png
Image is converted to a PNG before converting it
### raw
Image is converted to a raw file. It means that every pixel is a integer written as array.  
Example: On a 240x240 display with 16bit, it creates an array of 240*240*2=115'200 bytes.  
It is a good option if you use an Arduino with a TFT, since you can output directly the data read by this array without converting it.
### raw compressed
A simple compression method to group pixels with the same color.  
On Arduino, you can use the same TFT buffer and decompress it directly on the same buffer.  
A compressed RAW can be 20% to 60% smaller if the image is simple.
### raw zip
The raw image will be zipped.  You need this library on Arduino: https://github.com/bitbank2/unzipLIB to unzip it.  
A zipped RAW can be 50% to 90% smaller if the image is simple. But you need more time to unzip it.  

## RGB
As RAW, you can choose if you want a 8-bit pixel (RGB332) or 16-bit pixel (RGB565).

## Quality
As JPEG, you can choose the quality of the image (1-100%). Normally for a jpeg this value is from 90 to 95.

## Level
As Zipped RAW, you can choose a compression level (1 to 9). A simple compression is faster to decompress but need more space.


options.js
let Options = new class
{
  constructor()
  {
    this.file = null;
    this.div = null;
    this.image = null;
    this.divs = [];
    this.settings = {format:["hex", "dec", "bin"], prefix:"0x", suffix:"", pad:true, datatype:["uint8_t", "uint16_t", "uint32_t"], columns:16, skipbytes:0, asciicomments:false};
    this.imageSettings = {convertto:["none", "raw", "raw_compact", "raw_zip", "jpeg", "png"], rgb:["16bit", "8bit"], quality:95, level:2};
    this.typeSettings = {none:[], raw:["rgb"], raw_compact:["rgb"], raw_zip:["rgb", "level"], jpeg:["quality"], png:[]};
    this.inputs = [];
  }

  createDialog(f)
  {
    this.file = f;

    if(this.div != null) 
    {
        this.fillFileInfo();
        return;
    }

    this.div = _CN("div", {class:'dialog'}, null, document.body);
    this.divs['fileinfo'] = _CN("div", null, null, this.div);
    let butt = _CN("button", {}, ["convert"], this.div);
    this.divs['conversion'] = _CN("div", null, null, this.div);
    this.divs['image'] = _CN("div", {style:"display:none;"}, null, this.div);

    this.fillFileInfo();
    this.fillConversionSettings();
    this.fillImageConversionSettings();

    butt.addEventListener("click", ()=>{
      if(this.image != null && this.inputs["convertto"].value != "none")
      {
        Converter.convertImage(this.image, this.inputs);
      }
      else
      {
        Converter.convert(this.file, this.inputs);
      }
    });
  }

  fillFileInfo()
  {
    this.divs['fileinfo'].innerHTML = "";
    _CN("h2", null, ["File Info:"], this.divs['fileinfo']);
    _CN("span", null, [_CN("b",null, ["name: "]), this.file.name], this.divs['fileinfo']);
    _CN("span", null, [_CN("b",null, ["type: "]), this.file.type], this.divs['fileinfo']);
    _CN("span", null, [_CN("b",null, ["size: "]), this.file.size.toLocaleString() + " bytes"], this.divs['fileinfo']);
    createImageBitmap(this.file).then(im=> {
      console.log(im);
      this.image = _CN("canvas", {width:im.width, height:im.height});
      _CN("span", null, [
        this.image,
        " : " + this.image.width + "x" + this.image.height
      ], this.divs['fileinfo']);
      this.divs['image'].style.display = "inline-grid";
      let ctx = this.image.getContext("2d");
      ctx.drawImage(im,0,0,this.image.width,this.image.height,0,0,this.image.width,this.image.width);
    }).catch(ex=>{
      this.image = null;
      this.divs['image'].style.display = "none";
      console.log("not an image");
    });
  }

  fillConversionSettings()
  {
    this.divs['conversion'].innerHTML = "";
    let convTitle = _CN("h2", null, ["Conversion:", _CN("span", {class:"icon", style:"border-style:solid;border-width:1px;border-color:#00f;right:0px;"}, [" ❔"])], this.divs['conversion']);
    Object.keys(this.settings).forEach(k=>{
      let span = _CN("span", null, null, this.divs['conversion']);
      _CN("b", null, [k], span);
      this.inputs[k] = this.createInput(this.settings[k], span);
    });

    this.inputs["format"].addEventListener("change", ()=>{
      switch(this.inputs["format"].value)
      {
        case "hex": this.inputs["prefix"].value = "0x"; break;
        case "dec": this.inputs["prefix"].value = ""; break;
        case "bin": this.inputs["prefix"].value = "b"; break;
      }
    });

    convTitle.addEventListener("click", ()=>{
      Converter.showHelp("settings.md");
    });
  }

  fillImageConversionSettings()
  {
    this.divs['image'].innerHTML = "";
    let convTitle = _CN("h2", null, ["Image Settings:", _CN("span", {class:"icon", style:"border-style:solid;border-width:1px;border-color:#00f;right:0px;"}, [" ❔"])], this.divs['image']);
    Object.keys(this.imageSettings).forEach(k=>{
      let span = _CN("span", null, null, this.divs['image']);
      _CN("b", null, [k], span);
      this.inputs[k] = this.createInput(this.imageSettings[k], span);
      if(k != "convertto")
        this.inputs[k].style.visibility = "hidden";
    });

    this.inputs["convertto"].addEventListener("change", ()=>{
      var sels = this.divs['image'].getElementsByTagName("select");
      var inps = this.divs['image'].getElementsByTagName("input");
      var list = []
        .concat(Array.from(sels))
        .concat(Array.from(inps));
      for(var j=1;j<list.length;j++) list[j].style.visibility = "hidden";
      this.typeSettings[this.inputs["convertto"].value].forEach(k=>{this.inputs[k].style.visibility = "visible";});
    });

    convTitle.addEventListener("click", ()=>{
      Converter.showHelp("imagesettings.md");
    });
  }

  createInput(setting, parent)
  {
    let inp = null;
    if(Array.isArray(setting))
    {
      inp = _CN("select", null, null, parent);
      setting.forEach(o=>{_CN("option", {value:o}, [o], inp)});
    }
    else if(typeof setting == "boolean")
    {
      inp = _CN("input", {type:"checkbox"}, null, parent);
      if(setting) inp.setAttribute("checked", true);
    }
    else if(typeof setting == "number")
    {
      inp = _CN("input", {type:"number", value:setting}, null, parent);
    }
    else
    {
      inp = _CN("input", {type:"text", value:setting}, null, parent);
    }
    return inp;
  }
};